include/annotated.hpp

```diff
@@ -92,6 +93,74 @@ using DefaultSupportedTypes = SupportedTypes<>;
 static_assert(fair::meta::is_instantiation_of<DefaultSupportedTypes, SupportedTypes>);
 static_assert(fair::meta::is_instantiation_of<SupportedTypes<float, double>, SupportedTypes>);
 
+/**
+ * @brief Represents limits and optional validation for an Annotated<..> type.
+ *
+ * The `Limits` structure defines lower and upper bounds for a value of type `T`.
+ * Additionally, it allows for an optional custom validation function to be provided.
+ * This function should take a value of type `T` and return a `bool`, indicating
+ * whether the value passes the custom validation or not.
+ *
+ * Example:
+ * ```
+ * Annotated<float, "example float", Visible, Limits<0.f, 1024.f>>             exampleVar1;
+ * // or:
+ * constexpr auto isPowerOfTwo = [](const int &val) { return val > 0 && (val & (val - 1)) == 0; };
+ * Annotated<float, "example float", Visible, Limits<0.f, 1024.f, isPowerOfTwo>> exampleVar2;
+ * // or:
+ * Annotated<float, "example float", Visible, Limits<0.f, 1024.f, [](const int &val) { return val > 0 && (val & (val - 1)) == 0; }>> exampleVar2;
+ * ```
+ */
+template<auto LowerLimit, decltype(LowerLimit) UpperLimit, auto Validator = nullptr>
+    requires(requires(decltype(Validator) f, decltype(LowerLimit) v) {
+        { f(v) } -> std::same_as<bool>;
+    } || Validator == nullptr)
+struct Limits {
+    using ValueType = decltype(LowerLimit);
+    static constexpr ValueType MinRange = LowerLimit;
+    static constexpr ValueType MaxRange = UpperLimit;
+    static constexpr decltype(Validator) ValidatorFunc = Validator;
+
+    static constexpr bool
+    validate(const ValueType &value) noexcept {
+        if (LowerLimit == UpperLimit) {
+            if constexpr (Validator != nullptr) {
+                try {
+                    return Validator(value);
+                } catch (...) {
+                    return false;
+                }
+            } else {
+                return true; // if no validator and limits are same, return true by default
+            }
+        } else {
+            if constexpr (Validator != nullptr) {
+                try {
+                    if (value < LowerLimit || value > UpperLimit || !Validator(value)) return false;
+                } catch (...) {
+                    return false;
+                }
+            } else {
+                if (value < LowerLimit || value > UpperLimit) return false;
```

Comment:
Inclusive vs exclusive limits: I guess there's applications for both, I would have expected inclusive limits, e.g. I'd want to be able to set `gain = 0.0` while limiting it to [0.0, 1.0].

Reply:
That's the behaviour right now. The function returns false if it does not meet ... e.g. LowerLimit == 0.0 let's value == 0.0 pass.

Reply:
ah yes, one inversion too much for me, you are of course right, sorry for the noise.